### PR TITLE
kvm: set cpu topology only if cpucore per socket is set

### PR DESCRIPTION
**Pull request by ustcweizhou**
### Description

This PR fixes a regression issue in #4497 

In cloudstack 4.14 or before, the cpu topology is set only when cpucore per socket is set (to 4 or 6).
in other conditions, there is no cpu topology in vm xml definition.

with #4497, vm will have cpu topology in its xml definition, if cpucore per socket is not set.
```
    <topology sockets='<vm cpu cores>' cores='1' threads='1'/>
```
Not sure if it causes any issue. I think it would be better not to add this part in vm xml definition if cpucore per socket is not set.

fyi, in https://libvirt.org/formatdomain.html#cpu-model-and-topology, it says
```
topology
The topology element specifies requested topology of virtual CPU provided to the guest. Four attributes, sockets, dies, cores, and threads, accept non-zero positive integer values. They refer to the number of CPU sockets per NUMA node, number of dies per socket, number of cores per die, and number of threads per core, respectively. The dies attribute is optional and will default to 1 if omitted, while the other attributes are all mandatory. Hypervisors may require that the maximum number of vCPUs specified by the cpus element equals to the number of vcpus resulting from the topology.
```

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [X] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by ustcweizhou** [comment]
@rhtyd @DaanHoogland @sureshanaparti

**Review by rhtyd — APPROVED** [review]

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

**Review by rhtyd — APPROVED** [review]

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java [hunk 1]
```diff
@@ -4239,10 +4239,10 @@ private void setCpuTopology(CpuModeDef cmd, int vcpus, Map<String, String> detai
...
             } else if (vcpus % 4 == 0) {
                 numCoresPerSocket = 4;
-            } else {
-                numCoresPerSocket = 1;
             }
         }
-        cmd.setTopology(numCoresPerSocket, vcpus / numCoresPerSocket);
+        if (numCoresPerSocket > 0) {
+            cmd.setTopology(numCoresPerSocket, vcpus / numCoresPerSocket);
+        }
```
Comment (sureshanaparti) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java:4246]: code LGTM

**Review by sureshanaparti — APPROVED** [review]

**Comment by DaanHoogland** [comment]
@weizhouapache @rhtyd I see how this can bug a purist, but this is not really a critical bug is it? if so it is for milestone 4.14.1 . cc @sureshanaparti . Did i interpret my testing of #4497 wrong?

**Comment by weizhouapache** [comment]
> @weizhouapache @rhtyd I see how this can bug a purist, but this is not really a critical bug is it? if so it is for milestone 4.14.1 . cc @sureshanaparti . Did i interpret my testing of #4497 wrong?

@DaanHoogland 
let me explain what are the behaviors

Assume vm has 3 cpu cores, 

1. in older versions,

(1) by default, there is no cpu topology in vm definition
(2) if add vm setting "cpu.corespersocket" = 1, then in vm definition there should be 
```<topology sockets='3' cores='1' threads='1'/>```
but it is missing. 

2. #4497 fixes the issue in (2), but it also changes the default setting in (1). there are  ```<topology sockets='3' cores='1' threads='1'/>``` in both (1) and (2) above.

3. this pr reset the default setting in (1). there will be no cpu topology in vm definition, if cpu.corespersocket is not set.

I have no idea how critical this issue is.

**Comment by rhtyd** [comment]
Since the previous PR was already merged, let's merge the regression PR.

**Pull request merged**